src/Mocking.jl

```diff
@@ -34,125 +37,57 @@ function enable(; force::Bool=false)
-        end
-    end
-
-    # Generate the new method which will call the user's patch function. We need to perform
-    # this call instead of injecting the body expression to support closures.
-    sig, body = p.signature, p.body
-    params = call_parameters(sig)
-    push!(exprs, Expr(:(=), sig, Expr(:block, Expr(:call, body, params...))))
-
-    return Expr(:block, exprs...)
+    target::Callable
+    alternate::Function
 end
 
 macro patch(expr::Expr)
-    if expr.head == :function
-        name = expr.args[1].args[1]
-        params = expr.args[1].args[2:end]
-        body = expr.args[2]
-
-    # Short-form function syntax
-    elseif expr.head == :(=) && expr.args[1].head == :call
-        name = expr.args[1].args[1]
+    # Expect a named function in long-form or short-form
+    if expr.head === :function || (expr.head === :(=) && expr.args[1].head == :call)
+        target_name = expr.args[1].args[1]
         params = expr.args[1].args[2:end]
         body = expr.args[2]
-
-    # Anonymous function syntax
-    # elseif expr.head == :(->)
-        # TODO: Determine how this could be supported
+    elseif expr.head == :(->)
+        throw(ArgumentError("expression needs to be a named function"))
     else
         throw(ArgumentError("expression is not a function definition"))
     end
 
-    signature = Expr(:call, name, params...)
-
-    # Determine the bindings used in the signature
-    bindings = Bindings(signature)
+    patch_name = target_name isa Expr ? target_name.args[2].value : target_name
 
     # Need to evaluate the body of the function in the context of the `@patch` macro in
     # order to support closures.
     # func = Expr(:(->), Expr(:tuple, params...), body)
-    func = Expr(:(=), Expr(:call, gensym(), params...), body)
-
-    # Generate a translation between the external bindings and the runtime types and
-    # functions. The translation will be used to revise all bindings to be absolute.
-    translations = [Expr(:call, :(=>), QuoteNode(b), b) for b in bindings.external]
+    func = Expr(:(=), Expr(:call, gensym(patch_name), params...), body)
 
-    return esc(:(Mocking.Patch( $(QuoteNode(signature)), $func, Dict($(translations...)) )))
+    return esc(:(Mocking.Patch($target_name, $func)))
 end
 
 struct PatchEnv
-    mod::Module
+    mapping::Dict{Callable,Vector{Function}}
     debug::Bool
-
-    function PatchEnv(debug::Bool=false)
-        # Be careful not to call this code during pre-compilation otherwise we'll see the
-        # warning: "incremental compilation may be broken for this module"
-        m = Core.eval(Mocking, :(module $(gensym()) end))
-        new(m, debug)
-    end
 end
 
 function PatchEnv(patches::Array{Patch}, debug::Bool=false)
-    pe = PatchEnv(debug)
+    pe = PatchEnv(debug=debug)
     apply!(pe, patches)
     return pe
 end
 
 function PatchEnv(patch::Patch, debug::Bool=false)
-    pe = PatchEnv(debug)
+    pe = PatchEnv(debug=debug)
     apply!(pe, patch)
     return pe
 end
```

Comment:
these two methods are identical, it is kinda out of scope for this PR,  but could just drop the type constraints. and delete one.
(which would also make tuples etc work)

Reply:
I agree. I can adjust this.